### DISCUSSION TITLE
Is there a way to get much better results with CTC segmentation in French ?

### DISCUSSION THREAD

**Discussion by Ca-ressemble-a-du-fake** [NVIDIA-NeMo/NeMo#5839] [Q&A] [answered]
Hi,

First of all thanks for publishing all that pretrained models and writing detailed information on how TTS / ASR / Model ... work. These doc are very useful to me to learn AI.

I need to align text to audio to generate a French dataset for TTS training. So I followed [CTC segmentation] (https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/tools/CTC_Segmentation_Tutorial.ipyn) but the resulting segments are all off the track! 
I tried to shorten and widen WINDOW size, try to use different CTC models (stt_fr_conformer_ctc_large and stt_fr_citrinet_1024_gamma_0_25) with the same incorrect alignments (neither the beginning nor the end are correct).

Please note that for most segments the alignment confidence is < -2 (some are -0.05, ...) but even with those confidence no segment is correctly aligned with text.

If it works correctly with Spanish (see examples the Colab example) it should work with French, shouldn't it ?

So what can I do to get much better results ? Here is how I run the ctc segmentation :

```
python3 /home/caraduf/Nemo/NeMo/tools/ctc_segmentation/scripts/prepare_data.py \
--in_text=/home/caraduf/Tests/Nemo/input \
--output_dir=/home/caraduf/Tests/Nemo/output/processed/ \
--language='other' \
--cut_prefix=3 \
--model=stt_fr_conformer_ctc_large \
--audio_dir=/home/caraduf/Tests/Nemo/input
```
```
python3 /home/caraduf/Nemo/NeMo/tools/ctc_segmentation/scripts/run_ctc_segmentation.py \
--output_dir=/home/caraduf/Tests/Nemo/output \
--data=/home/caraduf/Tests/Nemo/output/processed \
--model=stt_fr_conformer_ctc_large \
--window_len=8000
```

Thanks in advance for your advice

**Comment by titu1994** [comment]
Like any trained model, alignment is rarely going to be perfect. Usually conformer models can give better alignment but since you have already tried that I'm not sure what else can be tried. 

Are you sure your text is punctuated and is not missing text that is spoken (things like uhm uh hmm if not present can all affect alignment calculation)

@ekmb any other advice ?

**Reply by Ca-ressemble-a-du-fake** [reply]
Thanks for your reply @titu1994. Original audio is clear with normal pace, no background noise, professional microphone... There is no "hum, well,..." maybe one repetition like "I, I have to admit that...".

Transcription is done by OpenAI whisper, is quite accurate, and includes punctuation.

The prepare_data script yields very good results with nearly one sentence per line.

Yet the first segment ends way too early in the middle ofna long sentence (it ends at 5 seconds instead of maybe 7 or 8 s but the confidence is -20). The other ones are completely off and some segments have nothing to do with their transcription although the confidence is above -2 (>-1). And I am quite estonished aligners do not work for me (I tried WhisperX, Whisperer) at all. I get better results by running Audacity silence finder and then running Whisper on each extract. I forgot to mention that stable_ts for Whisper also gives better results than the ones with segmentation but is still off by some #100ms (I did not measure the different offsets). 

If I run stt fr conformer transducer model for asr it gives very good results (unfortunately I could not find a way to input AudioSegment or byte array to the transcribe method as it can be done with Wav2vec2 or Whisper). So input audio seems to be in the expected format. (Wav file is pcm mono 16kHz).

I should try it in English and see if I can get good results with English contents so that I know if I am using it correctly.

Anyway if someone has a hint to make it better it would help me a lot 😊 !

**Reply by titu1994** [reply]
I dunno if Nemo forced aligned is going to give you accurate timestamps, but you can try it for CTC models from here - https://github.com/NVIDIA/NeMo/tree/main/tools/nemo_forced_aligner

The segmentation tool's in literature and NeMo don't support RNNT (yet) so it's forced alignments cannot be calculated (however if cer is 0, you can get greedy alignments of the RNNT model, but it will probably not be super accurate). 

Also, alignments of < 50 Ms is considered 
 excellent, there are very few model that can attain better alignment without explicitly training on a single file to minimize it's log likelihood explicitly.

**Reply by Ca-ressemble-a-du-fake** [reply]
Thank you. Haven't seen this NFA tool. Gonna try it just to be sure I don't reinvent the wheel! Are there paper or post or code that explain your last paragraph about " training on a single file to minimize its log likelihood" ? I cannot figure out the big picture of it :wink: !

**Reply by titu1994** [reply]
When you train a model, you reduce a loss over batched of samples to learn something. With alignment losses like CTC and RNNT you can overfit on a single example to get the alignment that is most probable.

**Reply by Ca-ressemble-a-du-fake** [reply]
But is it done "on-the-fly" for each single file ? If so where the ground-truth is taken from ? This is what I don't get when in the alignment process it occurs. Has it already been actually applied ?

**Reply by titu1994** [reply]
The ground truth must be provided. That's why it's forced alignment.

**Reply by titu1994** [reply]
Also, the most likely alignment doesn't necessarily mean it's the most correct acoustic alignments. Nothing in the loss specifies how it must align sequences, just to minimize the path to the most likely alignment. Likely to a loss function is not 1:1 alignment that humans may want.

**Reply by Ca-ressemble-a-du-fake** [reply]
I've just checked out the NFA on the first sentence and it was nearly perfect. Weird, right ? I used it with citrinet (as I did with ctc segmentation tool) and provided the sentence with punctuation (I don't know if the CTC segmentation tool actually uses  "with_punct", or "with_punctuation_normalized").

If I compare with ctc segmentation I had the following timestamps : "Start 0.07993884202453988  End 5.475810678680982" whereas for NFA I have something correct with "Start 0.08 End 6.80 (+ duration of 19.52)". While listening in Audacity it sounds perfect. 

But how to read the duration provided by the NFA tool ? The manual says "duration in samples" but it should be integers. For example I get for the starting word "0.08 0.08" and for the last word "6.80 19.52" .

Edit : quoting from the [CTC segmentation notebook](https://colab.research.google.com/github/NVIDIA/NeMo/blob/r1.0.0rc1/tutorials/tools/CTC_Segmentation_Tutorial.ipynb) 

> The `.txt` file **without punctuation** contains preprocessed text phrases that we're going to align within the audio file. Here, we split the text into sentences. Each line should contain a text snippet for alignment.

So this is why the CTC segmentation did not work for me, it should use the text with punctuation, shouldn't it ?

**Reply by titu1994** [reply]
NeMo forced aligner and CTC segmentation uses the models underlying tokenizer for alignment process - Nemo models are not trained with punctuation as of now (this will change soon), so you must reformat your text to have only lower case alphabet, and apostrophe and perhaps the minimal set of French vocab tokens. Only then will alignment will be successful. Otherwise you'll get values off by some duration. 

Another thing about Nemo forced aligned is that it's resolution is highly dependent on the stride of the model. Citrinet has 8x stride, so you'll get timestamps at 80ms intervals. Conformer has 4x stride so you'll get 40 ms offsets. You have to multiply the stride (4 or 8) by 10 ms for computing offset.

**Reply by Ca-ressemble-a-du-fake** [reply]
That's weird because I gave NFA text with punctuation and the result was awesome. I will poke around with those two tools now and more samples. I'll try to figure out what those durarions mean. Thanks again for your advices.

**Comment by titu1994** [comment]
Fyi @erastorgueva-nv maybe some questions here can be used to improve NFA

**Comment by ekmb** [comment, marked as answer]
Hi @Ca-ressemble-a-du-fake , CTC-segmentation takes care of punctuated text, processes and normalizes it. And makes so that for the alignment, only symbols supported by the model are used, so it shouldn't be an issue. I've tried fr_citrinet on a (the first audio from here)[https://librivox.org/compilation-de-poemes-012-by-various/], and it worked well.

A few questions:
- How fast is the speech in your audio?
- Have you checked the processed audio in `/home/caraduf/Tests/Nemo/output/processed/` after resampling --cut_prefix=3? does it sound ok? 

CTC-segmentation could struggle with start/end if the speaker talks too fast, but it is usually a few milliseconds issue, not a seconds diff. 
If you add `fr` (here)[https://github.com/NVIDIA/NeMo/blob/main/tools/ctc_segmentation/scripts/prepare_data.py#L50], it will run a basic text normalization, but even with "other" it should work.

**Reply by Ca-ressemble-a-du-fake** [reply]
Thank you @ekmb for your reply. The speech speed is normal (not fast at all). I did not listen to the resampled audio since it was already 16kHz but still this is a good idea, I will check it. I will also try the audio file you tested. The text seems correctly normalized (all lower case, no punctuation present, and the other files look alike with same punctuation in both of them). 

 I'll keep you posted !

**Reply by Ca-ressemble-a-du-fake** [reply]
@ekmb I tried the librivox example you took and ... silly me! I used `--cut-prefix=3` as shown in the [first tutorial](https://colab.research.google.com/github/NVIDIA/NeMo/blob/r1.0.0rc1/tutorials/tools/CTC_Segmentation_Tutorial.ipynb) that does not appear in the [updated tutorial](https://colab.research.google.com/github/NVIDIA/NeMo/blob/stable/tutorials/tools/CTC_Segmentation_Tutorial.ipynb).

That's why everything was shifted of 3 seconds! Now that I removed this option, the segmentation works great!

So now it is solved. Thank you for guiding me through this!
